tribits/examples/TribitsExampleProject2/packages/package1/CMakeLists.raw.cmake

```
@@ -0,0 +1,23 @@
+cmake_minimum_required(VERSION 3.23.0 FATAL_ERROR)
+
+if (COMMAND tribits_package)
+  message("Configuring raw CMake package Package1")
```

Should mention `TriBITS`, maybe
```suggestion
  message("Configuring raw CMake package Package1 with TriBITS")
```

---

Good suggestion.  I will create a follow-on PR for this.

Any more suggestions?  Did the documentation at:

* https://tribitspub.github.io/TriBITS/users_guide/index.html#how-to-implement-a-tribits-compliant-internal-package-using-raw-cmake
* https://tribitspub.github.io/TriBITS/users_guide/index.html#how-to-implement-a-tribits-compliant-external-package-using-raw-cmake

make sense?

---

> Any more suggestions? Did the documentation at:
> 
>     * [tribitspub.github.io/TriBITS/users_guide/index.html#how-to-implement-a-tribits-compliant-internal-package-using-raw-cmake](https://tribitspub.github.io/TriBITS/users_guide/index.html#how-to-implement-a-tribits-compliant-internal-package-using-raw-cmake)
> 
>     * [tribitspub.github.io/TriBITS/users_guide/index.html#how-to-implement-a-tribits-compliant-external-package-using-raw-cmake](https://tribitspub.github.io/TriBITS/users_guide/index.html#how-to-implement-a-tribits-compliant-external-package-using-raw-cmake)
> 
> 
> make sense?

Yes, I think it's pretty clear. I find
```
Provides CMake variables:
    <Package>_CONFIG or <Package>_TRIBITS_COMPLIANT_PACKAGE_CONFIG_FILE: Points to the file <Package>Config.cmake (i.e. ${CMAKE_CURRENT_LIST_FILE})
```
a little surprising/unusual and
```
The test <fullTestName> must not be added if the cache variable <fullTestName>_DISABLE is set to TRUE or if the cache variable <fullTestName>_SET_DISABLED_AND_MSG is set to non-empty (and the message string should be printed to STDOUT).
```
also feels pretty intrusive (but a package could just choose not to allow running tests if they are a subproject).

The management of `TPL`'s (e.g., Cuda) between external and internal raw CMake projects and TriBITS projects isn't discussed very much, though. Looking at https://github.com/TriBITSPub/TriBITS/tree/master/tribits/examples/TribitsExampleProject2/packages it seems that this should work as well, though. I guess we still need `FindTPL*` files for dependencies that are also used by the top-level `TriBITS` project, though?

---

> I find
> 
> >
> > Provides CMake variables:
> > * `<Package>_CONFIG` or `<Package>_TRIBITS_COMPLIANT_PACKAGE_CONFIG_FILE`: Points to the file `<Package>Config.cmake` (i.e. `${CMAKE_CURRENT_LIST_FILE}`)
> 
> a little surprising/unusual

It is because some of those `<Pakage>Config.cmake` files are included instead of found and if you don't call `find_package(<Package>)` then the var `<Package>_CONFIG` does not get set.  This is especially important for non-TriBITS-compliant external packages where TriBITS creates a `<buildDir>/external_packages/<Package>/<Package>Config.cmake` wrapper file that calls `find_package(<Package>)`.  This is mentioned in:

* https://tribitspub.github.io/TriBITS/build_ref/index.html#using-packages-from-the-build-tree-in-downstream-cmake-projects

And we don't want to set `<Package>_CONFIG` or `<Package>_DIR` because you don't want to interfere with CMake code that may be calling `find_package(<Package>)` that is expecting to find the non-TriBITS-compliant `<Package>Config.cmake` file or the `Find<Package>.cmake` file.  (This is an issue for HDF5, for example.)   Some of the issues with non-TriBITS-compliant external packages are explained in the section:

* https://tribitspub.github.io/TriBITS/users_guide/index.html#tricky-considerations-for-tribits-generated-tplname-config-cmake-files

but I need to mention the how the  `<Package>_CONFIG` and `<Package>_TRIBITS_COMPLIANT_PACKAGE_CONFIG_FILE` vars work in this system to make this work correctly.

I likely need to add another section to the TriBITS Users Guide that explains this system in a more complete way, pulling everything together. 

> and
>
> >The test `<fullTestName>` must not be added if the cache variable `<fullTestName>_DISABLE` is set to `TRUE` or if the cache variable `<fullTestName>_SET_DISABLED_AND_MSG` is set to non-empty (and the message string should be printed to STDOUT).
> 
> also feels pretty intrusive (but a package could just choose not to allow running tests if they are a subproject).

That is correct.  If a raw CMake package can't satisfy those requirements, they can just not turn off their tests.  Users of a TriBITS project expect to be able to disable any test by setting either of those variables.  Otherwise, it is very difficult to maintain the test suite across a bunch of build configurations and platforms.  For example, look at the tests being disabled in the Trilinos PR builds listed here:

* https://github.com/trilinos/Trilinos/blob/develop/packages/framework/ini-files/config-specs.ini

```
$ grep -n "_DISABLE " packages/framework/ini-files/config-specs.ini  | wc -l
450
```

But the truth is that there is a more to having TriBITS-compliant tests that just disabling tests based on those vars.  For example, to be submitted to CDash correctly, tests also need to set the label `<Package>` (I will add that to the requirements).  Also, some individual Kokkos unit tests get disabled on some build configurations like you see in:

* https://github.com/trilinos/Trilinos/blob/develop/packages/framework/ini-files/config-specs.ini

including:

```
$ grep -n "Kokkos_.*_EXTRA_ARGS" packages/framework/ini-files/config-specs.ini 
352:opt-set-cmake-var Kokkos_AlgorithmsUnitTest_MPI_1_EXTRA_ARGS STRING : --gtest_filter=-*Random_XorShift64:-*Random_XorShift1024
355:opt-set-cmake-var Kokkos_CoreUnitTest_Serial1_MPI_1_EXTRA_ARGS STRING : --gtest_filter=-*join_backward_compatibility
356:opt-set-cmake-var Kokkos_CoreUnitTest_Serial1_EXTRA_ARGS STRING : --gtest_filter=-*join_backward_compatibility
358:opt-set-cmake-var Kokkos_CoreUnitTest_Default_MPI_1_EXTRA_ARGS STRING : --gtest_filter=-*defaultdevicetype.shared_space
359:opt-set-cmake-var Kokkos_CoreUnitTest_Default_EXTRA_ARGS STRING : --gtest_filter=-*defaultdevicetype.shared_space
1320:opt-set-cmake-var Kokkos_CoreUnitTest_Cuda_MPI_1_EXTRA_ARGS STRING : "--gtest_filter=-cuda.debug_pin_um_to_host:cuda.debug_serial_execution"
```

The framework team and other developers need to ability to surgically disable tests without disabling everything.  (You don't want to be trusting a package on a given system if you can't run its test suite.  It sucks to have to debug issues caused by an upstream package in a downstream test suite.)

> The management of `TPL`'s (e.g., Cuda) between external and internal raw CMake projects and TriBITS projects isn't discussed very much, though.

In many ways, CUDA is no different than any other any other non-TriBITS-compliant external package/TPL.  You need to provide a `FindTPL<Package>.cmake` file along the lines documented in:

* https://tribitspub.github.io/TriBITS/users_guide/index.html#creating-findtpl-tplname-cmake-using-find-package-with-imported-targets

In the case of CUDA, `<tplName>` is CUDA and `<externalPkg>` is `CUDAToolkit`.  See:

https://github.com/TriBITSPub/TriBITS/blob/685c8d5fdb94fe591e4dbd1e0687c41943961461/tribits/core/std_tpls/FindTPLCUDA.cmake#L56-L60

> Looking at https://github.com/TriBITSPub/TriBITS/tree/master/tribits/examples/TribitsExampleProject2/packages it seems that this should work as well, though. I guess we still need `FindTPL*` files for dependencies that are also used by the top-level `TriBITS` project, though?

Yes, but it would be the responsibility of the top-level TriBITS project to provide those `FindTPL*.cmake` files, not the raw CMake packages that also use those external packages/TPLs.  For example, this is why CUDA dependencies work with the raw CMake Kokkos build system when building under the TriBITS project Trilinos.  Both `FindTPLCUDA.cmake` and the raw CMake Kokkos build system call `find_package(CUDAToolkit)` and we just need to make sure that both of those calls finds the same CUDA installation (which is not hard to ensure).

Thanks for looking all of this over!


I will be writing a SAND report that tries to explain the philosophy behind this system and why it offers something unique in productivity and build/test performance that you can't get with more course-grained packaging systems like Spack (but still providing the flexibility to build packages as individual CMake projects under Spack if you want).